packages/dds/tree/src/feature-libraries/treeTextCursor.ts

```diff
@@ -167,34 +141,180 @@ export class RootedTextCursor extends TextCursor {
                 parent: path,
-                parentIndex: this.indexStack[height],
-                parentField: this.keyStack[height],
+                parentIndex: this.getStackedNodeIndex(height),
+                parentField: key,
             };
         }
+
         path = {
             parent: path,
             parentIndex: this.index,
-            parentField: this.keyStack[length],
+            parentField: this.getStackedFieldKey(length - 1),
         };
         return path;
     }
+
+    public enterField(key: FieldKey): void {
+        // assert(this.mode === CursorLocationType.Nodes, "must be in nodes mode");
+        this.siblingStack.push(this.siblings);
+        this.indexStack.push(this.index);
+        this.index = -1;
+        this.siblings = key;
+    }
+
+    public get mode(): CursorLocationType {
+        return this.siblingStack.length % 2 === 0 ? CursorLocationType.Nodes : CursorLocationType.Fields;
+    }
+
+    public get pending(): boolean {
+        return false;
+    }
+
+    public nextField(skipPending: boolean): boolean {
+        if (this.mode === CursorLocationType.Nodes) {
+            // Implicitly enter first field
+            assert(this.index !== -1, "index should not be -1 in Nodes mode");
+            const fields = keys(this.getNode());
+            if (fields.length === 0) {
+                // No fields, so implicitly exit fields.
+                return false;
+            }
+
+            this.siblingStack.push(this.siblings);
+            this.siblings = fields;
+            this.indexStack.push(this.index);
+            this.index = 0;
+            return true;
+        }
+
+        // Already in "Fields" mode, so go to next field.
+        if (this.index === -1) {
+            // Navigated down to this field using a key, not iteration.
+            // Start iterating from this field:
+
+            // Iteration order is only defined for an individual iteration.
+            // If starting in the middle because accessed using `enterField`,
+            // we can unconditionally start iteration at the end.
+            this.upToNode();
+            return false;
+
+            // TODO: if requiring stable iteration order, use the below code:
+            // const key = this.siblings as FieldKey;
+            // const parents = this.siblingStack[this.siblingStack.length - 1] as readonly JsonableTree[];
+            // const parent: JsonableTree = parents[this.indexStack[this.indexStack.length - 1] as number];
+            // this.siblings = keys(parent);
+            // this.index = this.siblings.indexOf(key);
+            // // If key was not found, was in empty field, can start anywhere.
+            // this.index = 0;
+        }
+
+        assert(Array.isArray(this.siblings), "siblings should be an array");
+        this.index += 1;
+        if (this.index === this.siblings.length) {
+            this.upToNode();
+            return false;
+        }
+        return true;
+    }
```

Comment:
Note that this code is combining a couple of different methods (firstField, nextField) into a single modal method, which is at best perf neutral vs. having the caller directly invoke the method they need.

The ideal realworld-ish benchmark to measure the overhead of this convenience is probably to use the nextField API to scan the event id map in citm_catalog.json, counting entries that match some criteria, and compare with the original JsonCursor.

Reply:
I have benchmarks for this locally. I would like to merge this initial version before separately exploring the performance options around this in a follow up (hopefully a large perf test suite will land by then as well).

I have determined it makes a small perf difference, but this API makes loops easier to write.

A map style alternative is also in scope for exploring here.

I have added TODOs for these in the doc comments.